lib/streamlit/runtime/state/widgets.py

```diff
@@ -186,10 +186,58 @@ def register_widget(
     widget_id = _get_widget_id(element_type, element_proto, user_key)
     element_proto.id = widget_id
 
+    # Create the widget's updated metadata, and register it with session_state.
+    metadata = WidgetMetadata(
+        widget_id,
+        deserializer,
+        serializer,
+        value_type=ELEMENT_TYPE_TO_VALUE_TYPE[element_type],
+        callback=on_change_handler,
+        callback_args=args,
+        callback_kwargs=kwargs,
+    )
+    return register_widget_from_metadata(metadata, ctx, widget_func_name, element_type)
+
+
+def user_key_from_widget_id(wid: str) -> Optional[str]:
+    """Extract the user key used to generate a widget id, from that id.
+
+    Returns `None` instead of `"None"` if there was no user key,
+    for compatibility with the rest of the codebase, which represents it that way.
+
+    TODO This will not work correctly if the user actually provides "None"
+    as a key, but we can't avoid this kind of problem while storing the
+    string representation of the no-user-key sentinel as part of the widget id.
+    """
+    user_key = wid.split("-", maxsplit=2)[-1]
+    user_key = None if user_key == "None" else user_key
+    return user_key
```

Comment:
Are these caveats expressed in unit tests? "This will not work correctly" - what will happen if the user passes "None"? Should we detect that? Is there a way around this limitation (maybe by generating our widget keys differently?)

Reply:
We could get around it fully by making the key-present-or-not indicated by something like a wrapper type, or at least make it implausible that someone runs into it on accident by making the no-key sentinel something like "STREAMLIT_NO_USER_KEY" (although that contributes to user keys being annoying to read, which they already kinda are)